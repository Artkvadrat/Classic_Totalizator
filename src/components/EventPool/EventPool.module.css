.events {
  padding: 0;
  display: flex;
  flex-wrap: wrap;
  gap: 1em;
  list-style: none;
}
