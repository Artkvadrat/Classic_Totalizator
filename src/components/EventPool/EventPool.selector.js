import { createStructuredSelector } from 'reselect';
import parseDate from '../../services/dateParse/dateParse';

// Right now I use createStructuredSelector only for get needed structured data from response
// Need to create normal memoized selector when refactor the app. And when we will agree all contracts with backend

const isLoadingSelector = (state) => state.events.isLoading;
const eventsDataSelector = (state) =>
  state.events.eventsData
    .sort((a, b) => Date.parse(a.startTime) - Date.parse(b.startTime))
    .map(
      ({
        id,
        startTime,
        participant1,
        participant2,
        sportName,
        margin,
        possibleResults,
        isEnded
      }) => ({
        key: id,
        date: parseDate(startTime),
        player1: participant1.name,
        player2: participant2.name,
        sport: sportName,
        margin: `${margin}%`,
        possibleResults,
        isEnded
      })
    );

const eventPoolSelector = createStructuredSelector({
  isLoading: isLoadingSelector,
  eventsData: eventsDataSelector
});

export default eventPoolSelector;
