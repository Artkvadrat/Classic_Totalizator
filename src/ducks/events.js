import HTTPService from '../services/HTTPService';

const REQUESTED = 'events/requested';
const RECEIVED = 'events/received';
const FINISH = 'event/finish';

const requested = () => ({
  type: REQUESTED
});

const received = (data) => ({
  type: RECEIVED,
  payload: data
});

const finish = (data) => ({
  type: FINISH,
  payload: data
});

export const finishEvent = (data) => (dispatch) => {
  HTTPService.request({
    method: 'PATCH',
    path: '/api/Events/finishEvent',
    body: data
  });

  dispatch(finish(data));
};

export const getEvents = () => (dispatch) => {
  dispatch(requested());

  HTTPService.request({ path: '/api/Events/feed' }).then((data) => {
    dispatch(received(data));
  });
};

const initialState = {
  isLoading: false,
  eventsData: []
};

const eventsReducer = (state = initialState, action) => {
  switch (action.type) {
    case REQUESTED:
      return {
        ...state,
        isLoading: true
      };

    case RECEIVED:
      return {
        ...state,
        isLoading: false,
        eventsData: [...action.payload.events]
      };

    case FINISH:
      return {
        ...state,
        eventsData: state.eventsData.map((event) => {
          if (event.id === action.payload.id) {
            return { ...event, isEnded: true };
          }

          return event;
        })
      };

    default:
      return state;
  }
};

export default eventsReducer;
