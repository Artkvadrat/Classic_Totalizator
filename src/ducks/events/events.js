import HTTPService from '../../services/HTTPService/HTTPService';

export const REQUESTED = 'events/requested';
export const RECEIVED = 'events/received';
export const FINISH = 'event/finish';

const requested = () => ({
  type: REQUESTED
});

const received = (data) => ({
  type: RECEIVED,
  payload: data
});

const finish = (data) => ({
  type: FINISH,
  payload: data
});

export const finishEvent = (data) => (dispatch) =>
  HTTPService.request({
    method: 'POST',
    path: '/api/Events/finish',
    body: data
  }).then(dispatch(finish(data)));

export const getEvents = () => (dispatch) => {
  dispatch(requested());

  return HTTPService.request({ path: '/api/Events' }).then((data) => {
    dispatch(received(data));
  });
};

const initialState = {
  isLoading: false,
  eventsData: []
};

const eventsReducer = (state = initialState, action) => {
  switch (action.type) {
    case REQUESTED:
      return {
        ...state,
        isLoading: true
      };

    case RECEIVED:
      return {
        ...state,
        isLoading: false,
        eventsData: [...action.payload.events]
      };

    case FINISH:
      return {
        ...state,
        eventsData: state.eventsData.map((event) => {
          if (event.id === action.payload.id) {
            return { ...event, isEnded: true };
          }

          return event;
        })
      };

    default:
      return state;
  }
};

export default eventsReducer;
