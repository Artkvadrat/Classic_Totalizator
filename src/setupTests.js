import enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

enzyme.configure({ adapter: new Adapter() });

const localStorageMock = {
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn()
};
window.localStorage = localStorageMock;
