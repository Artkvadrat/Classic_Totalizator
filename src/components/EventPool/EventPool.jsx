import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Skeleton, Table, Row, Col, Statistic } from 'antd';
import { Link } from 'react-router-dom';

import { getEvents } from '../../ducks/events/events';
import EventFinish from '../EventFinish/EventFinish';

const EventPool = () => {
  const dispatch = useDispatch();
  const { isLoading, eventsData } = useSelector((state) => state.events);

  useEffect(() => {
    dispatch(getEvents());
  }, [dispatch]);

  return isLoading ? (
    <Skeleton active />
  ) : (
    <>
      <Row justify="end" style={{ margin: 16 }}>
        <Col span={20}>
          <h1>All events</h1>
        </Col>
        <Col span={2}>
          <Statistic title="Total" value={eventsData.length} />
        </Col>
        <Col span={2}>
          <Statistic
            title="Unfinished"
            value={eventsData.filter((el) => !el.isEnded).length}
          />
        </Col>
      </Row>

      <Table dataSource={eventsData}>
        <Table.Column title="Time" dataIndex="date" key="date" />
        <Table.Column title="Sport" dataIndex="sport" key="sport" />
        <Table.Column title="Player 1" dataIndex="player1" key="player1" />
        <Table.Column title="Player 2" dataIndex="player2" key="player2" />
        <Table.Column title="Margin" dataIndex="margin" key="margin" />
        <Table.Column
          title="Event Status"
          key="status"
          defaultFilteredValue={['Not Resolved']}
          filters={[{ text: 'Not Resolved', value: 'Not Resolved' }]}
          onFilter={(_, record) => !record.isEnded}
          render={({ isEnded, possibleResults, key }) =>
            isEnded ? (
              'Event finished'
            ) : (
              <EventFinish id={key} possibleResults={possibleResults} />
            )
          }
        />
        <Table.Column
          title="Action"
          key="action"
          render={({ isEnded, key }) =>
            !isEnded && <Link to={`/edit/${key}`}>Edit</Link>
          }
        />
      </Table>
    </>
  );
};

export default EventPool;
