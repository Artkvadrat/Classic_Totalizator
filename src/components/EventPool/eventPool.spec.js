import React from 'react';
import { shallow, mount } from 'enzyme';
import { useDispatch, useSelector } from 'react-redux';
import { Skeleton, Table, Statistic } from 'antd';
import { Link } from 'react-router-dom';

import { getEvents } from '../../ducks/events/events';

import '../../setupTests';

import EventPool from './EventPool';

const { Column } = Table;

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn()
}));

jest.mock('../../ducks/events/events', () => ({
  getEvents: jest.fn()
}));

describe('EventPool component', () => {
  const dispatch = jest.fn();

  beforeEach(() => {
    useDispatch.mockReturnValue(dispatch);
    getEvents.mockReturnValue(Symbol.for('events'));

    useSelector.mockReturnValue({
      isLoading: true,
      eventsData: []
    });
  });

  it('should dispatch action on mount', () => {
    expect(dispatch).not.toHaveBeenCalled();
    expect(getEvents).not.toHaveBeenCalled();

    mount(<EventPool />);

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith(Symbol.for('events'));
    expect(getEvents).toHaveBeenCalledTimes(1);
  });

  it('should render skeleton if loading in progress', () => {
    const result = shallow(<EventPool />);
    const skeleton = result.find(Skeleton);

    expect(skeleton.length).toEqual(1);
    expect(skeleton.prop('active')).toEqual(true);

    expect(result).toMatchSnapshot();
  });

  it('should render results with no data', () => {
    useSelector.mockReturnValue({
      isLoading: false,
      eventsData: []
    });

    const result = shallow(<EventPool />);

    expect(result.find(Statistic).length).toEqual(2);
    result.find(Statistic).forEach((item) => {
      expect(item.prop('value')).toEqual(0);
    });

    expect(result).toMatchSnapshot();
  });

  describe('should render results with data', () => {
    useSelector.mockReturnValue({
      isLoading: false,
      eventsData: [
        {
          key: '593aa557-5af8-472d-ac3d-fe674ce7da67',
          date: '20:13 22.03',
          player1: 'Hammes - Bode',
          player2: 'Gwen Greenholt',
          sport: 'Joany_Green',
          margin: '2%',
          possibleResults: ['W1', 'X', 'W2'],
          isEnded: false
        },
        {
          key: '40dca2cb-af06-4711-b03b-830387a7a6e9',
          date: '12:36 23.03',
          player1: 'Beier Group',
          player2: 'Klocko - Champlin',
          sport: 'UFC',
          margin: '5.1%',
          possibleResults: ['W1', 'W2'],
          isEnded: true
        },
        {
          key: '868495b6-5d41-4209-9087-5720af12e352',
          date: '03:01 25.03',
          player1: 'Herman Barton',
          player2: 'Gwen Greenholt',
          sport: 'UFC',
          margin: '25%',
          possibleResults: ['W1', 'X', 'W2'],
          isEnded: false
        },
        {
          key: 'ce1c4820-f3f5-4332-87e0-58cc6a757972',
          date: '13:21 25.03',
          player1: 'string',
          player2: 'Kulas Inc',
          sport: 'football',
          margin: '1.5%',
          possibleResults: ['W1', 'X', 'W2'],
          isEnded: true
        },
        {
          key: 'e774d7eb-d373-4291-b1c4-813487b25bcf',
          date: '08:29 26.03',
          player1: 'Kathryn Effertz',
          player2: 'Lyle Russel',
          sport: 'UFC',
          margin: '3%',
          possibleResults: ['W1', 'X', 'W2'],
          isEnded: false
        }
      ]
    });

    const result = shallow(<EventPool />);

    it('should render statistics of events', () => {
      expect(result.find(Statistic).length).toEqual(2);
      expect(result.find(Statistic).map((el) => el.prop('value'))).toEqual([
        5,
        3
      ]);
    });

    it('should render table of events', () => {
      const table = result.find(Table);
      expect(table.length).toEqual(1);
    });

    it('table column status should render correct logic', () => {
      const statusColumn = result.find(Column).at(5);
      const renderFunc = statusColumn.prop('render');
      expect(renderFunc({ isEnded: true })).toEqual('Event finished');
      expect(
        renderFunc({ isEnded: false, possibleResults: ['X'], key: 'id' })
      ).toBeDefined();

      const filterFunc = statusColumn.prop('onFilter');
      expect(filterFunc('', { isEnded: false })).toEqual(true);
      expect(filterFunc('', { isEnded: true })).toEqual(false);
    });

    it('table column action should render correct logic', () => {
      const statusColumn = result.find(Column).at(6);

      const renderFunc = statusColumn.prop('render');
      expect(renderFunc({ isEnded: true, key: 'pageID' })).toEqual(false);
      expect(renderFunc({ isEnded: false, key: 'pageID' })).toEqual(
        <Link to="/edit/pageID">Edit</Link>
      );
    });

    expect(result).toMatchSnapshot();
  });
});
