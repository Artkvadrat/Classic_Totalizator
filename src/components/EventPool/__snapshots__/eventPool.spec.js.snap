// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<Fragment>
  <Row
    justify="end"
    style={
      Object {
        "margin": 16,
      }
    }
  >
    <Col
      span={20}
    >
      <h1>
        All events
      </h1>
    </Col>
    <Col
      span={2}
    >
      <withConfigConsumer(Statistic)
        title="Total"
        value={5}
      />
    </Col>
    <Col
      span={2}
    >
      <withConfigConsumer(Statistic)
        title="Unfinished"
        value={3}
      />
    </Col>
  </Row>
  <Table
    dataSource={
      Array [
        Object {
          "date": "20:13 22.03",
          "isEnded": false,
          "key": "593aa557-5af8-472d-ac3d-fe674ce7da67",
          "margin": "2%",
          "player1": "Hammes - Bode",
          "player2": "Gwen Greenholt",
          "possibleResults": Array [
            "W1",
            "X",
            "W2",
          ],
          "sport": "Joany_Green",
        },
        Object {
          "date": "12:36 23.03",
          "isEnded": true,
          "key": "40dca2cb-af06-4711-b03b-830387a7a6e9",
          "margin": "5.1%",
          "player1": "Beier Group",
          "player2": "Klocko - Champlin",
          "possibleResults": Array [
            "W1",
            "W2",
          ],
          "sport": "UFC",
        },
        Object {
          "date": "03:01 25.03",
          "isEnded": false,
          "key": "868495b6-5d41-4209-9087-5720af12e352",
          "margin": "25%",
          "player1": "Herman Barton",
          "player2": "Gwen Greenholt",
          "possibleResults": Array [
            "W1",
            "X",
            "W2",
          ],
          "sport": "UFC",
        },
        Object {
          "date": "13:21 25.03",
          "isEnded": true,
          "key": "ce1c4820-f3f5-4332-87e0-58cc6a757972",
          "margin": "1.5%",
          "player1": "string",
          "player2": "Kulas Inc",
          "possibleResults": Array [
            "W1",
            "X",
            "W2",
          ],
          "sport": "football",
        },
        Object {
          "date": "08:29 26.03",
          "isEnded": false,
          "key": "e774d7eb-d373-4291-b1c4-813487b25bcf",
          "margin": "3%",
          "player1": "Kathryn Effertz",
          "player2": "Lyle Russel",
          "possibleResults": Array [
            "W1",
            "X",
            "W2",
          ],
          "sport": "UFC",
        },
      ]
    }
    rowKey="key"
  >
    <Column
      dataIndex="date"
      key="date"
      title="Time"
    />
    <Column
      dataIndex="sport"
      key="sport"
      title="Sport"
    />
    <Column
      dataIndex="player1"
      key="player1"
      title="Player 1"
    />
    <Column
      dataIndex="player2"
      key="player2"
      title="Player 2"
    />
    <Column
      dataIndex="margin"
      key="margin"
      title="Margin"
    />
    <Column
      filters={
        Array [
          Object {
            "text": "Not Resolved",
            "value": "Not Resolved",
          },
        ]
      }
      key="status"
      onFilter={[Function]}
      render={[Function]}
      title="Event Status"
    />
    <Column
      key="action"
      render={[Function]}
      title="Action"
    />
  </Table>
</Fragment>
`;

exports[`EventPool component should render results with no data 1`] = `
<Fragment>
  <Row
    justify="end"
    style={
      Object {
        "margin": 16,
      }
    }
  >
    <Col
      span={20}
    >
      <h1>
        All events
      </h1>
    </Col>
    <Col
      span={2}
    >
      <withConfigConsumer(Statistic)
        title="Total"
        value={0}
      />
    </Col>
    <Col
      span={2}
    >
      <withConfigConsumer(Statistic)
        title="Unfinished"
        value={0}
      />
    </Col>
  </Row>
  <Table
    dataSource={Array []}
    rowKey="key"
  >
    <Column
      dataIndex="date"
      key="date"
      title="Time"
    />
    <Column
      dataIndex="sport"
      key="sport"
      title="Sport"
    />
    <Column
      dataIndex="player1"
      key="player1"
      title="Player 1"
    />
    <Column
      dataIndex="player2"
      key="player2"
      title="Player 2"
    />
    <Column
      dataIndex="margin"
      key="margin"
      title="Margin"
    />
    <Column
      filters={
        Array [
          Object {
            "text": "Not Resolved",
            "value": "Not Resolved",
          },
        ]
      }
      key="status"
      onFilter={[Function]}
      render={[Function]}
      title="Event Status"
    />
    <Column
      key="action"
      render={[Function]}
      title="Action"
    />
  </Table>
</Fragment>
`;

exports[`EventPool component should render skeleton if loading in progress 1`] = `
<Skeleton
  active={true}
  avatar={false}
  paragraph={true}
  title={true}
/>
`;
